docs/rfc/block-registration.md

```diff
@@ -0,0 +1,523 @@
+		}
+	}
+}
+```
+
+Attributes provide the structured data needs of a block. They can exist in different forms when they are serialized, but they are declared together under a common interface.
+
+See the [the attributes documentation](/docs/designers-developers/developers/block-api/block-attributes.md) for more details.
+
+### Style Variations
+
+*   Type: `array`
+*   Optional
+*   Localized: Yes (`label`)
+*   Property: `styles`
+
+```json
+{ 
+	"styles": [ 
+		{ "name": "default", "label": "Default", "isDefault": true }, 
+		{ "name": "other", "label": "Other" }
+	]
+}
+```
+
+Block styles can be used to provide alternative styles to block. It works by adding a class name to the block's wrapper. Using CSS, a theme developer can target the class name for the style variation if it is selected.
+
+Plugins and Themes can also register [custom block style](/docs/designers-developers/developers/filters/block-filters.md#block-style-variations) for existing blocks.
+
+### Editor Script
+
+* Type: `string|object` ([WPDefinedAsset](#WPDefinedAsset))
+* Optional
+* Localized: No
+* Property: `editorScript`
+
+```json
+{ "editorScript": "file:build/editor.asset.json" }
+```
+
+Block type editor script definition. It will only be enqueued in the context of the editor.
+
+### Script
+
+* Type: `string|object` ([WPDefinedAsset](#WPDefinedAsset))
+* Optional
+* Localized: No
+* Property: `script`
+
+```json
+{
+	"script": {
+		"handle": "my-plugin-notice",
+		"file": "index.js",
+		"dependencies": [ "wp-element", "wp-components" ],
+		"version": "3.0.0"
+	}
+}
+```
+
+Block type frontend script definition. It will be enqueued both in the editor and when viewing the content on the front of the site.
+
+### Editor Style
+
+* Type: `string|object` ([WPDefinedAsset](#WPDefinedAsset))
+* Optional
+* Localized: No
+* Property: `editorStyle`
+
+```json
+{
+	"editorStyle": {
+		"handle": "my-plugin-notice-editor",
+		"file": "editor.css",
+		"dependencies": [ "wp-edit-blocks" ]
+	}
+}
+```
+
+Block type editor style definition. It will only be enqueued in the context of the editor.
+
+### Style
+
+* Type: `string|object` ([WPDefinedAsset](#WPDefinedAsset))
+* Optional
+* Localized: No
+* Property: `style`
+
+```json
+{ "style": "my-plugin-notice" }
+```
+
+Block type frontend style definition. It will be enqueued both in the editor and when viewing the content on the front of the site.
+
+### Render Callback
+
+* Type: `string` ([WPDefinedPropertyFile](#WPDefinedPropertyFile))
+* Optional
+* Localized: No
+* Property: `renderCallback`
+
+```json
+{ "renderCallback": "file:my-render-callback.php" }
+```
+
+This is a pointer to a php file returning a render callback php function. The render callback is function called when the block is rendered on the frontend. It's used to generate the frontend markup dynamically.
+
+See the [dynamic blocks documentation](/docs/designers-developers/developers/tutorials/block-tutorial/creating-dynamic-blocks.md) for more details.
+
```

Comment:
Could there be another property of supported post types? It will likely be the case they developers will want to limit where blocks are displayable once the widget areas are blocks as well.

Reply:
It's out of the scope of this proposal but I'm sure this is something which is going to be discussed soon as we start to see blocks in the widget areas.

Reply:
But this is already an issue, as I for want to limit blocks to post type for custom post types. Take for example, a product post type, you would likely want to limit this to just product based blocks. 

My worry here is when existing blocks are converted use this format, we if add more fields to the registered, that will require developers to go back and rework all their blocks again.

Reply:
There is an existing PHP filter which:

> Filters the allowed block types for the editor, defaulting to true (all block types supported).

https://github.com/WordPress/wordpress-develop/blob/8081adfefa61dd58dbbebe107aa9a6068f865c8b/src/wp-admin/edit-form-blocks.php#L138-L148

It allows applying a different set of block per post type, id or anything else that a post object contains. I'm sure it's not perfect and it would be great to have more control over and maybe we could extend `parent` field to use it as a way to limit block's exposure to the list of CPTs. It's defintely something we should further explore but I consider it as a new feature. We should discuss it separately. There is also a related issue which impacts `InnerBlocks` where you can't filter `allowedBlocks` option when used inside `edit` function (#15682). It seems like those 2 items could be tackled together.